File: ts/src/gameloop/queue.ts
import { Socket } from '../server/socket.js';
import { Message } from '../server/message.js';

export interface QueueMessage {
    from: number;
    message: Message;
}

export class Queue {
    private messages: QueueMessage[] = [];
    private stopped = false;

    start(s0: Socket, s1: Socket): void {
        s0.onMessage = (msg) => {
            if (this.stopped) return;
            this.messages.push({ from: 1, message: msg });
        };

        s1.onMessage = (msg) => {
            if (this.stopped) return;
            this.messages.push({ from: 2, message: msg });
        };
    }

    stop(): void {
        this.stopped = true;
    }

    flush(): QueueMessage[] | null {
        if (this.messages.length === 0) {
            return null;
        }
        const messages = this.messages;
        this.messages = [];
        return messages;
    }
}

File: java/src/main/java/com/hornflakes/gorunsrestwalk/gameloop/Queue.java
package com.hornflakes.gorunsrestwalk.gameloop;

import com.hornflakes.gorunsrestwalk.server.Message;
import com.hornflakes.gorunsrestwalk.server.Socket;

import java.util.ArrayList;
import java.util.List;
import java.util.concurrent.ConcurrentLinkedQueue;

public final class Queue {

    public record QueueMessage(int from, Message message) {}

    private final ConcurrentLinkedQueue<QueueMessage> messages = new ConcurrentLinkedQueue<>();
    private Thread reader0;
    private Thread reader1;

    public void start(Socket s0, Socket s1) {
        reader0 = Thread.startVirtualThread(() -> readLoop(s0, 1));
        reader1 = Thread.startVirtualThread(() -> readLoop(s1, 2));
    }

    public void stop() {
        if (reader0 != null) reader0.interrupt();
        if (reader1 != null) reader1.interrupt();
    }

    public List<QueueMessage> flush() {
        var batch = new ArrayList<QueueMessage>();
        QueueMessage msg;
        while ((msg = messages.poll()) != null) {
            batch.add(msg);
        }
        return batch.isEmpty() ? null : batch;
    }

    private void readLoop(Socket socket, int from) {
        try {
            while (!Thread.currentThread().isInterrupted()) {
                var msg = socket.in().take();
                messages.add(new QueueMessage(from, msg));
            }
        } catch (InterruptedException e) {
            // stopped
        }
    }
}
